# Spacetime Intervals

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/lookbusy1344/Relativity/blob/main/Python/Spacetime%20interval.ipynb)

## In classical physics, distance is independent of time

In classical physics, the distance between two points in space is given by the Pythagorean theorem, and is independent of time. The distance $d$ between two points in a three-dimensional space with coordinates $(x_1, y_1, z_1)$ and $(x_2, y_2, z_2)$ is given by:

$$d = \sqrt{\Delta x^2 + \Delta y^2 + \Delta z^2}$$

For example, if we have two points in space with coordinates $(1, 2, 3)$ and $(4, 5, 6)$, the distance between them is:

$$d = \sqrt{(4 - 1)^2 + (5 - 2)^2 + (6 - 3)^2} = \sqrt{3^2 + 3^2 + 3^2} = \sqrt{27} = 3\sqrt{3} \approx 5.196$$

## In special relativity, space and time are intertwined

In *special relativity*, distances and time can vary according to the observer's frame of reference. Two events might be 1 metre apart from one frame, but 10 metres apart in a different frame. To account for this, we need a new way of measuring distances (intervals) that incorporates both space and time. *Spacetime intervals* combine both spatial and temporal dimensions and do not depend on the observer's motion. Two events will always have the same spacetime interval, even though the spacial and temporal distances may differ for different observers.

The *spacetime interval squared* $\Delta s^2$ between two events is:

$$
\Delta s^2 = c^2 \Delta t^2 - \Delta x^2 - \Delta y^2 - \Delta z^2
$$

where $c$ is the speed of light, $\Delta t$ is the time difference, and $\Delta x$, $\Delta y$, $\Delta z$ are the spatial differences.

3 types of spacetime intervals can be defined based on the sign of $\Delta s^2$: *time-like*, *space-like*, and *light-like*. More about this later.

## Making it simpler, 1-D space

To simplify the math, instead of 3 spacial dimensions ($x, y, z$) and time, we can use only 1 spatial dimension ($x$) and time ($t$). All spacial coordinates appear on a 1-D line. The spacetime interval can then be expressed as:

$$\Delta s^2 = c^2 \Delta t^2 - \Delta x^2$$

## Example Calculation

In [1]:
import math


def spacetime_interval_1d(t1: float, x1: float, t2: float, x2: float) -> float:
    """
    Return s^2 for two events in 1+1D Minkowski space (c=1).
    """
    dt = t2 - t1  # Time difference
    dx = x2 - x1  # Space difference
    return dt**2 - dx**2


def lorentz_transform_1d(t: float, x: float, v: float) -> tuple[float, float]:
    """
    Apply the Lorentz transformation for 1+1D spacetime.

    Returns:
    tuple: (t', x') in the moving frame
    """
    gamma = 1 / math.sqrt(1 - v**2)
    t_prime = gamma * (t - v * x)
    x_prime = gamma * (x - v * t)
    return t_prime, x_prime


# Simultaneous events, spatially separated by 4 units
# in the rest frame (t1, x1) = (0, 0) and (t2, x2) = (0, 4).
t1, x1 = 0.0, 0.0
t2, x2 = 0.0, 4.0

interval_s = spacetime_interval_1d(t1, x1, t2, x2)

# Moving frame with v = 0.8 (80% of the speed of light)
v = 0.8
t1p, x1p = lorentz_transform_1d(t1, x1, v)
t2p, x2p = lorentz_transform_1d(t2, x2, v)
interval_s_prime = spacetime_interval_1d(t1p, x1p, t2p, x2p)

print(
    f"Rest frame  : Δt = {t2 - t1}, Δx = {x2 - x1}, interval squared = {interval_s:.3f}"
)
print(
    f"Moving frame: Δt = {t2p - t1p:.3f}, Δx = {x2p - x1p:.3f}, interval squared = {interval_s_prime:.3f}"
)

Rest frame  : Δt = 0.0, Δx = 4.0, interval squared = -16.000
Moving frame: Δt = -5.333, Δx = 6.667, interval squared = -16.000


In this example:

- In the rest frame, the events are simultaneous (Δt = 0), spatially separated by 4 units (Δx = 4).
- In the moving frame, the events are no longer simultaneous (Δt = -5.3), and the spatial separation has expanded (Δx = 6.67).
- In both frames, the interval squared is -16. Spacetime intervals do not change between inertial frames.

So what does an interval of -16 mean?

## 3 types of interval (+, −, 0)

Depending on the sign, there are 3 types of interval:

| Interval Type         | $\Delta s^2$ Value | Explanation                                                                                                                                                                                                                  |
|-----------------------|:------------------:|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Time-like             | $> 0$              | A slower-than-light signal could connect the events. There always exists a frame in which the two events occur at the **same location** but at different times.                                                              |
| Space-like            | $< 0$              | No signal, even light, could connect the events. There exists a frame in which the two events are **simultaneous**, but occur at different locations.                                                                        |
| Light-like (null)     | $= 0$              | Events lie on the path of a light signal. In every frame, the events occur at different locations and times, but are always connected at **light speed**.                                                                    |

*Time-like* and *light-like* intervals can be causally connected, while *space-like* intervals cannot.

So -16 is a *space-like* interval. In any possible frame the events are spatially separated and cannot be causally connected.

Units here are arbitrary to keep the math simple. We will use familiar metres and seconds later.

### Change in simultaneity

Notice that two simultaneous events in one frame are not simultaneous in another! They have also moved physically further apart. Space and time have in some sense been "smeared out" into each other, strongly suggesting that the two are interconnected, hence relativity uses the term 'spacetime'.

### Invariant intervals

All this changing of distances and times is confusing! But even though the space and time components change, the spacetime interval squared remains fixed (−16), which is the key result. This is how we measure a unchanging separation between to points in spacetime.

## Second example

Now lets use everyday units of metres and seconds. Lets take two events separated by 5 seconds and 3 million kilometers in the rest frame (frame A). This time we use my `relativity_lib.py` library for increased accuracy and convenience. Its included in the repository.

*3 million km is approximately 10 light-seconds, about 7 times further away than the moon. We use a large distance so the effect is more pronounced. Light moves so quickly it covers 300m in a microsecond!*

The second frame (frame B) is moving at 0.99c (99% light speed) compared to the first frame. This is about 297,000 km/s.

In [2]:
# Download relativity_lib.py from your GitHub repo if not present. Important for colab.
import os

if not os.path.exists("relativity_lib.py"):
    !wget -q https://raw.githubusercontent.com/lookbusy1344/Relativity/main/Python/relativity_lib.py

import mpmath as mp
import relativity_lib as rl

rl.configure(50)
x1, t1 = mp.mpf(0), mp.mpf(0)  # origin 0,0
x2 = x1 + mp.mpf("3000000000")  # 3 million km, approx 10 light-seconds away
t2 = t1 + mp.mpf(5)  # 5 seconds later

# Frame A - at rest with respect to the events
interval_s = rl.spacetime_interval_1d((t1, x1), (t2, x2))
print("Frame A - at rest with respect to the events")
print(f"Separation {t2} seconds in time and {x2 / mp.mpf(1000)} km apart")
print(f"Interval squared before transform: {rl.format_mpf(interval_s, 1)} m^2")

# Lorentz transform for frame B moving at 99% of the speed of light
vel = mp.mpf("0.99")  # 99% of the speed of light
t_prime, x_prime = rl.lorentz_transform_1d(t2, x2, vel * rl.c)

# Frame B info
print()
print(f"Frame B - Lorentz transform with velocity v = {vel}c")
interval_s_prime = rl.spacetime_interval_1d((t1, x1), (t_prime, x_prime))
print(
    f"Separation {rl.format_mpf(t_prime)} seconds in time and {rl.format_mpf(x_prime / mp.mpf(1000), 0)} km apart"
)
print(f"Interval squared after transform: {rl.format_mpf(interval_s_prime, 1)} m^2")

Frame A - at rest with respect to the events
Separation 5.0 seconds in time and 3000000.0 km apart
Interval squared before transform: -6,753,112,053,157,955,900.0 m^2

Frame B - Lorentz transform with velocity v = 0.99c
Separation -34.78 seconds in time and 10,746,832 km apart
Interval squared after transform: -6,753,112,053,157,955,900.0 m^2


## Explanation

To understand how this calculation works, lets look how the relative velocity interacts with the $x$ and $t$ coordinates, changing the time and space separations between the two events.

### Relative speed - Lorentz factor

Frame B is moving at 0.99c compared to frame A. First we calculate the Lorentz factor ($\textcolor{orange}{\gamma}$) for this relative velocity:

$$\textcolor{orange}{\gamma} = \frac{1}{\sqrt{1 - 0.99^2}} \approx \textcolor{orange}{7.09}$$

The Lorentz factor determines how much time and space are distorted between the two frames (read more about [time dilation and length contraction](https://github.com/lookbusy1344/Relativity/blob/main/Python/Time%20dilation.ipynb)). One second in frame A takes $1 \times 7.09 = 7.09$ seconds in frame B, and 1 metre in frame A shrinks to $1 / 7.09 \approx 0.14$ m in frame B.

### New spatial and temporal separations

Now lets calculate the time and space separations for frame B. Spacial coordinate $\textcolor{red}{\Delta x}$ is red, time coordinate $\textcolor{green}{\Delta t}$ is green. The Lorentz factor $\textcolor{orange}{\gamma}$ is in orange. Notice how they depend on each other:

$$\Delta t’ = \textcolor{orange}{\gamma}(\textcolor{green}{\Delta t} - \frac{v \textcolor{red}{\Delta x}}{c^2})$$

$$\Delta x’ = \textcolor{orange}{\gamma}(\textcolor{red}{\Delta x} - v \textcolor{green}{\Delta t})$$

### Frame B time separation

Filling in the values, the time difference in frame B is:

$$ \Delta t' = \textcolor{orange}{\gamma} \left( \textcolor{green}{\Delta t} - \frac{v \textcolor{red}{\Delta x}}{c^2} \right) = \textcolor{orange}{7.09} \left( \textcolor{green}{5} - \frac{0.99c \times (\textcolor{red}{3{\times}10^9})}{c^2} \right) \approx -34.78 $$

Its changed from 5 seconds in frame A to approximately 34.78 seconds in frame B. Its negative so the second event happens before the first in frame B.

### Frame B spatial separation

Meanwhile the spatial separation in frame B is:

$$ \Delta x' = \textcolor{orange}{\gamma} \left( \textcolor{red}{\Delta x} - v \textcolor{green}{\Delta t} \right) = \textcolor{orange}{7.09} \left( \textcolor{red}{3{\times}10^9} - 0.99c \times \textcolor{green}{5} \right) \approx 10{,}746{,}832{,}000 $$

Its changed from 3 million km in frame A to approximately 10.75 million km in frame B.

## What does the interval squared mean?

Spacetime intervals squared here are measured in $m^2$, what does that mean? How can a spacetime distance be measured in square metres?!

As we saw above, intervals can be *time-like*, *space-like* or *light-like*. *Time-like* intervals can be causally connected, and are positive. *Space-like* intervals cannot be causally connected, and are negative. *Light-like* intervals equal zero, and are the separation of two events connected by light.

Let’s take as an example an interval squared of $100 m^2$. It’s positive so the interval is *time-like*. There is at least one inertial frame where both events occur in the same location in space, and are separated in time by:

$$
\frac{\sqrt{100m^2}}{c} \approx 0.000000033 s
$$

Which is about 33 nanoseconds. However in different frames the spacial and temporal separations will be different (only the spacetime interval remains constant).

Back to our code, where the separation is $-6{,}753{,}112{,}053{,}157{,}955{,}900 m^2$. This is negative so it’s *space-like*, the two events cannot be causally connected. Nothing can travel from one event to the other regardless of the reference frame chosen (not even light).

There is however an inertial frame where the events occur simultaneously, and the distance between them in that frame is:

$$
\sqrt{6{,}753{,}112{,}053{,}157{,}955{,}900 m^2} \approx 2{,}598{,}675{,}057 m
$$

Note we remove the minus sign before performing the square root.

So we know the following about the two events:
- Frame A, stationary, where they are 5 seconds and 3,000,000 km apart
- Frame B, moving 0.99c, where they are 34.8 seconds and 10,746,832 km apart
- In a third frame the events occur simultaneously and are 2,598,675 km apart
- The interval is space-like, the events cannot be causally connected, in any frame

Lets confirm this calculation in Python:

In [3]:
distance_str = rl.min_separation_str(interval_s_prime)
# type, time, distance = rl.min_separation(interval_s_prime)
print(f"Minimum separation: {distance_str}")

Minimum separation: space-like 2,598,675,057.24 m


## Conclusion

Instead of fixed separations in space and time separately, in special relativity we have fixed *spacetime intervals*, fulfilling the same role, while individual space and time components can change depending on the observer's frame of reference and relative velocity.

More about my [special relativity tools here](https://github.com/lookbusy1344/Relativity/blob/main/Python/README.md)

More about spacetime intervals: https://en.wikipedia.org/wiki/Spacetime#Spacetime_interval